File: src/main/java/waterpower/common/block/turbines/TileEntityTurbine.java
package waterpower.common.block.turbines;

import net.minecraft.block.state.IBlockState;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fluids.FluidRegistry;
import waterpower.client.gui.DefaultGuiIds;
import waterpower.common.block.reservoir.Reservoir;
import waterpower.common.block.reservoir.TileEntityReservoir;
import waterpower.common.block.tileentity.TileEntityRotor;
import waterpower.util.WPLog;

public class TileEntityTurbine extends TileEntityRotor {

    public static final int MAX_OUTPUT = 32767;

    public TileEntityTurbine() {
        super();
    }
    
    public TileEntityTurbine(TurbineType type) {
    	super(type.percent, 10000000);
    	temporaryType = type;
    }
    
    @Override
    public IBlockState getBlockState() {
    	return super.getBlockState().withProperty(BlockTurbine.TURBINE_TYPES, getType());
    }
    
    @Override
    protected void func_190201_b(World p_190201_1_) {
    	super.func_190201_b(p_190201_1_);
    	worldObj = p_190201_1_;
    	getType();
    }
    
    TurbineType temporaryType = null;
    
    public TurbineType getType() {
    	if (temporaryType != null) return temporaryType;
    	temporaryType = TurbineType.values()[getBlockMetadata()];
        production = temporaryType.percent;
        maxStorage = 10000000;
        return temporaryType;
    }

    private TileEntityReservoir getWater(World world, BlockPos pos) {
        TileEntityReservoir reservoir = null;
        BlockPos newPos = pos.offset(getFacing().getOpposite());
        if (newPos.getY() != pos.getY())
        	return null;
        if (Reservoir.isRes(world, newPos))
            reservoir = (TileEntityReservoir) world.getTileEntity(newPos);
        return reservoir;
    }

    public double getWater() {
        TileEntityReservoir pair = getWater(worldObj, getPos());
        return pair.getFluidAmount();
    }

    @Override
    protected double computeOutput(World world, BlockPos pos) {
        TileEntityReservoir pair = getWater(world, pos);
        if (pair == null || pair.getFluidTank() == null || pair.getFluidStackfromTank() == null || pair.getFluidfromTank() == null)
            return 0;
        else {
            WPLog.debugLog("fluidAmount=" + pair.getFluidAmount());
            WPLog.debugLog("maxUse=" + pair.getType().maxUse);
            double use = Math.min(pair.getFluidAmount(), pair.getType().maxUse);
            WPLog.debugLog("use=" + use);
            int multiple = 0;
            if (pair.getFluidfromTank() == FluidRegistry.WATER)
                multiple = 1;
            else if (FluidRegistry.isFluidRegistered("steam") && pair.getFluidfromTank() == FluidRegistry.getFluid("steam")
                    || FluidRegistry.isFluidRegistered("ic2steam") && pair.getFluidfromTank() == FluidRegistry.getFluid("ic2steam"))
                multiple = 5;
            if (multiple == 0)
                return 0;
            double baseEnergy = use * getType().percent / 2048 * multiple;
            WPLog.debugLog("baseEnergy" + baseEnergy);
            // WPLog.debugLog("speed" + speed);
            double per = tickRotor();
            WPLog.debugLog("per=" + per);
            if (per > 0) {
                double energy = baseEnergy * per; // * ((double) speed / 50);
                WPLog.debugLog("energy = " + energy);
                if (energy > 0) {
                    pair.useLiquid((int) use);
                    damageRotor(1);
                }
                WPLog.debugLog("use=" + use);
                return energy;
            }
        }
        return 0;
    }

    @Override
    public String getName() {
        return getType().getShowedName();
    }

    @Override
    public int getGuiId() {
        return DefaultGuiIds.get("tileEntityTurbine");
    }

    @Override
    public boolean isActive() {
        return super.isActive() && (storage < maxStorage);
    }

    @Override
    protected boolean allowedSendPacketTank() {
        return false;
    }

}

File: src/main/kotlin/waterpower/common/block/turbine/TileEntityTurbine.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.turbine

import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.fluids.FluidRegistry
import waterpower.annotations.HasGui
import waterpower.annotations.Register
import waterpower.common.block.container.ContainerRotor
import waterpower.common.block.reservoir.TileEntityReservoir
import waterpower.common.block.tile.TileEntityRotorGenerator
import waterpower.common.init.WPBlocks

@HasGui(guiClass = GuiTurbine::class, containerClass = ContainerRotor::class)
open class TileEntityTurbine(val type: Turbines) : TileEntityRotorGenerator(type.getOutput(), 10000000.0, type.getTier()) {

    override fun getName() = type.getLocalizedName()

    override fun getBlockState() = super.getBlockState().withProperty(WPBlocks.turbine.TYPES, type)!!

    private fun getWater(world: World, pos: BlockPos): TileEntityReservoir? {
        var reservoir: TileEntityReservoir? = null
        val newPos = pos.offset(getFacing().opposite)
        if (newPos.y != pos.y)
            return null
        val te = world.getTileEntity(newPos)
        if (te is TileEntityReservoir)
            reservoir = te
        return reservoir
    }

    override fun computeOutput(world: World, pos: BlockPos): Double {
        val pair = getWater(world, pos)
        if (pair?.getFluidTank()?.fluid?.fluid == null)
            return 0.0
        else {
            val fluidStack = pair.getFluidTank().fluid!!
            val use = minOf(pair.getFluidTank().fluidAmount, pair.type.maxExtract)
            var multiple = 0
            if (fluidStack.fluid == FluidRegistry.WATER)
                multiple = 1
            else if (FluidRegistry.isFluidRegistered("steam") && fluidStack.fluid == FluidRegistry.getFluid("steam") || FluidRegistry.isFluidRegistered("ic2steam") && fluidStack.fluid == FluidRegistry.getFluid("ic2steam"))
                multiple = 5
            if (multiple == 0)
                return 0.0
            val baseEnergy = 1.0 * use / 2048 * type.getOutput() * multiple
            val per = getEfficiency()
            if (per > 0) {
                val energy = baseEnergy * per
                if (energy > 0) {
                    pair.useLiquid(use)
                    damageRotor(1)
                }
                return energy
            }
        }
        return 0.0
    }
}

@Register("waterpower.turbine.mk1")
class TileEntityTurbineMK1 : TileEntityTurbine(Turbines.MK1)

@Register("waterpower.turbine.mk2")
class TileEntityTurbineMK2 : TileEntityTurbine(Turbines.MK2)

@Register("waterpower.turbine.mk3")
class TileEntityTurbineMK3 : TileEntityTurbine(Turbines.MK3)

@Register("waterpower.turbine.mk4")
class TileEntityTurbineMK4 : TileEntityTurbine(Turbines.MK4)

@Register("waterpower.turbine.mk5")
class TileEntityTurbineMK5 : TileEntityTurbine(Turbines.MK5)

@Register("waterpower.turbine.mk6")
class TileEntityTurbineMK6 : TileEntityTurbine(Turbines.MK6)

@Register("waterpower.turbine.mk7")
class TileEntityTurbineMK7 : TileEntityTurbine(Turbines.MK7)

@Register("waterpower.turbine.mk8")
class TileEntityTurbineMK8 : TileEntityTurbine(Turbines.MK8)

@Register("waterpower.turbine.mk9")
class TileEntityTurbineMK9 : TileEntityTurbine(Turbines.MK9)

@Register("waterpower.turbine.mk10")
class TileEntityTurbineMK10 : TileEntityTurbine(Turbines.MK10)

@Register("waterpower.turbine.mk11")
class TileEntityTurbineMK11 : TileEntityTurbine(Turbines.MK11)

@Register("waterpower.turbine.mk12")
class TileEntityTurbineMK12 : TileEntityTurbine(Turbines.MK12)